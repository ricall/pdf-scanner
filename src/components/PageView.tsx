import React, { SyntheticEvent, useState } from 'react';
import { Paper } from "@material-ui/core";
import { CloseReason, OpenReason, SpeedDial, SpeedDialAction, SpeedDialIcon } from "@material-ui/lab";
import { ArrowBack, ArrowForward, Delete, Rotate90DegreesCcw } from "@material-ui/icons";
import { createStyles, makeStyles, Theme } from "@material-ui/core/styles";
import { Page } from "react-pdf";
import { PdfPage } from "../types";

const useStyles = makeStyles((theme: Theme) => createStyles({
    paper: {
        width: 270,
        height: 380,
    },
    speedDial: {
        position: 'absolute',
        '&.MuiSpeedDial-directionLeft': {
            bottom: theme.spacing(2),
            right: theme.spacing(2)
        },
    },
    control: {
        padding: theme.spacing(2),
    },
}));

export interface PageProps {
    index: number,
    page: PdfPage,
    selected?: boolean,
    onClick: (page: number) => void,
    onRotatePage: (page: number) => void,
    onMovePageRight: (page: number) => void,
    onMovePageLeft: (page: number) => void,
    onDeletePage: (page: number) => void,
}

export const PageView = ({
    index,
    page,
    selected,
    onClick,
    onRotatePage,
    onMovePageRight,
    onMovePageLeft,
    onDeletePage
}: PageProps) => {
    const classes = useStyles();

    const [ open, setOpen ] = useState(false);

    const handleOpen = (e: SyntheticEvent<any>, _: OpenReason) => {
        e.stopPropagation();
        setOpen(true);
    }
    const handleClose = (e: SyntheticEvent<any>, _: CloseReason) => {
        e.stopPropagation();
        setOpen(false);
    }
    const handleAction = (action: (page: number) => void) => (e: SyntheticEvent<any>) => {
        e.stopPropagation();
        action(index);
    }

    return (
        <Paper elevation={selected ? 5 : 1} className={classes.paper}>
            <Page
                width={270}
                height={380}
                pageNumber={page.page + 1}
                rotate={page.angle}
                onClick={handleAction(onClick)}>
                {selected && <SpeedDial
                  ariaLabel="Actions"
                  className={classes.speedDial}
                  open={open}
                  icon={<SpeedDialIcon/>}
                  direction="left"
                  onOpen={handleOpen}
                  onClose={handleClose}>
                  <SpeedDialAction
                    icon={<Rotate90DegreesCcw/>}
                    tooltipTitle="Rotate page"
                    onClick={handleAction(onRotatePage)}/>
                  <SpeedDialAction
                    icon={<ArrowForward/>}
                    tooltipTitle="Move page right"
                    onClick={handleAction(onMovePageRight)}/>
                  <SpeedDialAction
                    icon={<ArrowBack/>}
                    tooltipTitle="Move page left"
                    onClick={handleAction(onMovePageLeft)}/>
                  <SpeedDialAction
                    icon={<Delete/>}
                    tooltipTitle="Delete page"
                    onClick={handleAction(onDeletePage)}/>
                </SpeedDial>
                }
            </Page>
        </Paper>
    );
}