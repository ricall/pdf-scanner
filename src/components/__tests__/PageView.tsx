import React from 'react';
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import { PageView } from "../PageView";

describe('PageView component', () => {

    it('can be rendered', () => {
        const wrapper = shallow(<PageView
            index={1}
            page={{
                page: 1,
                angle: 180
            }}
            selected={true}
            onClick={jest.fn()}
            onRotatePage={jest.fn()}
            onMovePageRight={jest.fn()}
            onMovePageLeft={jest.fn()}
            onDeletePage={jest.fn()}
        />);
        const tree = toJson(wrapper);
        expect(tree).toMatchSnapshot();
    });

});
