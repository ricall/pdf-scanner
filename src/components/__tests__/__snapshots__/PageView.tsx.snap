// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageView component can be rendered 1`] = `
<WithStyles(ForwardRef(Paper))
  className="makeStyles-paper-1"
  elevation={5}
>
  <ForwardRef(Page)
    height={380}
    onClick={[Function]}
    pageNumber={2}
    rotate={180}
    width={270}
  >
    <WithStyles(ForwardRef(SpeedDial))
      ariaLabel="Actions"
      className="makeStyles-speedDial-2"
      direction="left"
      icon={<WithStyles(ForwardRef(SpeedDialIcon)) />}
      onClose={[Function]}
      onOpen={[Function]}
      open={false}
    >
      <WithStyles(ForwardRef(SpeedDialAction))
        icon={<Memo />}
        onClick={[Function]}
        tooltipTitle="Rotate page"
      />
      <WithStyles(ForwardRef(SpeedDialAction))
        icon={<Memo />}
        onClick={[Function]}
        tooltipTitle="Move page right"
      />
      <WithStyles(ForwardRef(SpeedDialAction))
        icon={<Memo />}
        onClick={[Function]}
        tooltipTitle="Move page left"
      />
      <WithStyles(ForwardRef(SpeedDialAction))
        icon={<Memo />}
        onClick={[Function]}
        tooltipTitle="Delete page"
      />
    </WithStyles(ForwardRef(SpeedDial))>
  </ForwardRef(Page)>
</WithStyles(ForwardRef(Paper))>
`;
